bigquery/tests/unit/test_dataset.py

```diff
@@ -186,11 +186,22 @@ def test_from_string(self):
         self.assertEqual(got.project, "string-project")
         self.assertEqual(got.dataset_id, "string_dataset")
 
+    def test_from_string_w_prefix(self):
+        cls = self._get_target_class()
+        got = cls.from_string("prefix:string-project.string_dataset")
+        self.assertEqual(got.project, "prefix:string-project")
+        self.assertEqual(got.dataset_id, "string_dataset")
+
     def test_from_string_legacy_string(self):
         cls = self._get_target_class()
         with self.assertRaises(ValueError):
             cls.from_string("string-project:string_dataset")
 
+    def test_from_string_w_incorrect_prefix(self):
```

Comment:
Let's add an additional test where the project ID / dataset ID contains an illegal `.` character. Another way to say that, is the string contains too many "parts". e.g. `google.com:project-id.dataset_id.table_id`. This should also fail with `ValueError`.